File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <iostream>
#include <vector>

using namespace std;

int dfs(const vector<int>& numbers, int target, int index, int current_sum) {
    // 모든 숫자를 다 사용했을 때
    if (index == numbers.size()) {
        return (current_sum == target) ? 1 : 0;
    }
    
    // 현재 숫자를 더하는 경우와 빼는 경우를 탐색
    return dfs(numbers, target, index + 1, current_sum + numbers[index]) + 
           dfs(numbers, target, index + 1, current_sum - numbers[index]);
}

int solution(vector<int> numbers, int target) {
    return dfs(numbers, target, 0, 0);
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = [0]  # 리스트는 내부에서 수정 가능

    def dfs(index, current_sum):
        if index == len(numbers):
            if current_sum == target:
                answer[0] += 1
            return
        dfs(index + 1, current_sum + numbers[index])
        dfs(index + 1, current_sum - numbers[index])
    
    dfs(0, 0)
    return answer[0]
